File: src/main/kotlin/com/iodigital/weather/WeatherService.java
package com.iodigital.weather;

import com.iodigital.weather.api.WeatherAPIClient;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;

@Service
public class WeatherService {
    private final WeatherAPIClient api;
    private final WeatherRepository repository;

    public WeatherService(final WeatherAPIClient api, final WeatherRepository repository) {
        this.api = api;
        this.repository = repository;
    }

    public Flux<WeatherInfo> getAll() {
        return repository.findAll();
    }

    public Flux<WeatherInfo> getForCity(final String city) {
        return repository
            .findAllByCityIgnoreCase(city)
            .switchIfEmpty(
                api
                    .getWeather(city)
                    .flatMapMany(apiResponse -> repository.saveAll(apiResponse.toWeatherInfoList()))
            );
    }
}

File: src/main/kotlin/com/iodigital/weather/WeatherService.kt
package com.iodigital.weather

import com.iodigital.weather.api.WeatherAPIClient
import kotlinx.coroutines.flow.toList
import org.springframework.stereotype.Service

@Service
class WeatherService(
    private val api: WeatherAPIClient,
    private val repository: WeatherRepository
) {
    suspend fun getAll(): List<WeatherInfo> =
        repository.findAll().toList()

    suspend fun getForCity(city: String): List<WeatherInfo> =
        repository
            .findAllByCityIgnoreCase(city)
            .toList()
            .takeIf { it.isNotEmpty() }
            ?: api
                .getWeather(city)
                .let { repository.saveAll(it.toWeatherInfoList()).toList() }
}
